File: packages/progress/src/cusum.ts
import type { SessionEvent, RecalibrationResolution } from './types'
import { CUSUM_SLACK_FACTOR, CUSUM_THRESHOLD_FACTOR } from './config'

export interface CUSUMResult {
  breakpoints: number[]
  upperAccumulator: number[]
  lowerAccumulator: number[]
}

export interface RegimeShiftResult {
  breakpoints: number[]
  promptNeeded: boolean
  cusumState: { upper: number; lower: number }
}

export function runCUSUM(
  paceRatios: number[],
  referenceMean: number,
  std: number,
): CUSUMResult {
  const n = paceRatios.length
  if (n < 3) {
    return {
      breakpoints: [],
      upperAccumulator: new Array(n).fill(0),
      lowerAccumulator: new Array(n).fill(0),
    }
  }

  const effectiveStd = std < 1e-6 ? 0.05 : std
  const k = CUSUM_SLACK_FACTOR * effectiveStd
  const h = CUSUM_THRESHOLD_FACTOR * effectiveStd

  const upperAccumulator: number[] = new Array(n)
  const lowerAccumulator: number[] = new Array(n)
  const breakpoints: number[] = []

  let currentRef = referenceMean

  for (let i = 0; i < n; i++) {
    const z = paceRatios[i] - currentRef
    const prevUpper = i > 0 ? upperAccumulator[i - 1] : 0
    const prevLower = i > 0 ? lowerAccumulator[i - 1] : 0

    upperAccumulator[i] = Math.max(0, prevUpper + z - k)
    lowerAccumulator[i] = Math.min(0, prevLower + z + k)

    if (upperAccumulator[i] > h || lowerAccumulator[i] < -h) {
      breakpoints.push(i)
      upperAccumulator[i] = 0
      lowerAccumulator[i] = 0
      const start = Math.max(0, i - 4)
      const recent = paceRatios.slice(start, i + 1)
      currentRef = recent.reduce((s, r) => s + r, 0) / recent.length
    }
  }

  return { breakpoints, upperAccumulator, lowerAccumulator }
}

export function detectRegimeShifts(
  sessions: SessionEvent[],
  exceptionalIds: Set<string>,
  posteriorMean: number,
  resolutions: RecalibrationResolution[],
): RegimeShiftResult {
  const activeSessions = sessions.filter(
    (s) =>
      s.source === 'active' &&
      s.plannedMinutes != null &&
      s.plannedMinutes > 0 &&
      s.activeMinutes != null &&
      s.activeMinutes > 0 &&
      (!s.sessionId || !exceptionalIds.has(s.sessionId)),
  )

  if (activeSessions.length < 3) {
    return {
      breakpoints: [],
      promptNeeded: false,
      cusumState: { upper: 0, lower: 0 },
    }
  }

  const paceRatios = activeSessions.map(
    (s) => s.activeMinutes! / s.plannedMinutes!,
  )

  const mean = paceRatios.reduce((s, r) => s + r, 0) / paceRatios.length
  const sumSq = paceRatios.reduce((s, r) => s + (r - mean) ** 2, 0)
  const std = Math.sqrt(sumSq / (paceRatios.length - 1))

  let referenceMean = posteriorMean

  // Handle recalibration resolutions
  const lastResolution = resolutions.length > 0
    ? resolutions[resolutions.length - 1]
    : null

  if (lastResolution) {
    if (lastResolution.resolution === 'acknowledged') {
      const resolvedTime = new Date(lastResolution.resolvedAt).getTime()
      const sessionsAfter = activeSessions.filter(
        (s) => new Date(s.date).getTime() > resolvedTime,
      )
      if (sessionsAfter.length > 0) {
        const afterRatios = sessionsAfter.map(
          (s) => s.activeMinutes! / s.plannedMinutes!,
        )
        referenceMean =
          afterRatios.reduce((s, r) => s + r, 0) / afterRatios.length
      }
    }
  }

  const result = runCUSUM(paceRatios, referenceMean, std)

  // Determine if prompt is needed:
  // prompt is needed if CUSUM is currently in alarm state (not just historical breakpoints)
  // and no unresolved resolution exists after the last breakpoint
  const lastBreakpoint = result.breakpoints.length > 0
    ? result.breakpoints[result.breakpoints.length - 1]
    : null

  let promptNeeded = false
  if (lastBreakpoint !== null) {
    if (!lastResolution) {
      promptNeeded = true
    } else {
      const resolvedTime = new Date(lastResolution.resolvedAt).getTime()
      const breakpointSession = activeSessions[lastBreakpoint]
      if (breakpointSession) {
        const breakpointTime = new Date(breakpointSession.date).getTime()
        promptNeeded = breakpointTime > resolvedTime
      }
    }
  }

  const n = paceRatios.length
  return {
    breakpoints: result.breakpoints,
    promptNeeded,
    cusumState: {
      upper: n > 0 ? result.upperAccumulator[n - 1] : 0,
      lower: n > 0 ? result.lowerAccumulator[n - 1] : 0,
    },
  }
}

File: college/mydeliverables/phase1-ESA/for-submit/code/packages/py-progress/src/py_progress/cusum.py
from __future__ import annotations

from datetime import datetime

from py_progress.config import CUSUM_SLACK_FACTOR, CUSUM_THRESHOLD_FACTOR
from py_progress.dates import date_to_ms
from py_progress.types import (
    CUSUMResult,
    RecalibrationResolution,
    RegimeShiftResult,
    SessionEvent,
)


def run_cusum(
    pace_ratios: list[float],
    reference_mean: float,
    std: float,
) -> CUSUMResult:
    n = len(pace_ratios)
    if n < 3:
        return CUSUMResult(
            breakpoints=[],
            upperAccumulator=[0.0] * n,
            lowerAccumulator=[0.0] * n,
        )

    effective_std = 0.05 if std < 1e-6 else std
    k = CUSUM_SLACK_FACTOR * effective_std
    h = CUSUM_THRESHOLD_FACTOR * effective_std

    upper_accumulator: list[float] = [0.0] * n
    lower_accumulator: list[float] = [0.0] * n
    breakpoints: list[int] = []
    current_ref = reference_mean

    for i in range(n):
        z = pace_ratios[i] - current_ref
        prev_upper = upper_accumulator[i - 1] if i > 0 else 0.0
        prev_lower = lower_accumulator[i - 1] if i > 0 else 0.0

        upper_accumulator[i] = max(0.0, prev_upper + z - k)
        lower_accumulator[i] = min(0.0, prev_lower + z + k)

        if upper_accumulator[i] > h or lower_accumulator[i] < -h:
            breakpoints.append(i)
            upper_accumulator[i] = 0.0
            lower_accumulator[i] = 0.0
            start = max(0, i - 4)
            recent = pace_ratios[start : i + 1]
            current_ref = sum(recent) / len(recent)

    return CUSUMResult(
        breakpoints=breakpoints,
        upperAccumulator=upper_accumulator,
        lowerAccumulator=lower_accumulator,
    )


def detect_regime_shifts(
    sessions: list[SessionEvent],
    exceptional_ids: set[str],
    posterior_mean: float,
    resolutions: list[RecalibrationResolution],
) -> RegimeShiftResult:
    active_sessions = [
        s
        for s in sessions
        if s.get("source") == "active"
        and s.get("plannedMinutes") is not None
        and s["plannedMinutes"] > 0
        and s.get("activeMinutes") is not None
        and s["activeMinutes"] > 0
        and (not s.get("sessionId") or s["sessionId"] not in exceptional_ids)
    ]

    if len(active_sessions) < 3:
        return RegimeShiftResult(
            breakpoints=[],
            promptNeeded=False,
            cusumState={"upper": 0.0, "lower": 0.0},
        )

    pace_ratios = [s["activeMinutes"] / s["plannedMinutes"] for s in active_sessions]

    mean = sum(pace_ratios) / len(pace_ratios)
    sum_sq = sum((r - mean) ** 2 for r in pace_ratios)
    std = (sum_sq / (len(pace_ratios) - 1)) ** 0.5

    reference_mean = posterior_mean

    last_resolution = resolutions[-1] if resolutions else None

    if last_resolution and last_resolution["resolution"] == "acknowledged":
        resolved_time = datetime.fromisoformat(
            last_resolution["resolvedAt"].replace("Z", "+00:00")
        ).timestamp() * 1000
        sessions_after = [
            s for s in active_sessions if date_to_ms(s["date"]) > resolved_time
        ]
        if sessions_after:
            after_ratios = [
                s["activeMinutes"] / s["plannedMinutes"] for s in sessions_after
            ]
            reference_mean = sum(after_ratios) / len(after_ratios)

    result = run_cusum(pace_ratios, reference_mean, std)

    last_breakpoint = result.breakpoints[-1] if result.breakpoints else None

    prompt_needed = False
    if last_breakpoint is not None:
        if not last_resolution:
            prompt_needed = True
        else:
            resolved_time = datetime.fromisoformat(
                last_resolution["resolvedAt"].replace("Z", "+00:00")
            ).timestamp() * 1000
            breakpoint_session = active_sessions[last_breakpoint]
            breakpoint_time = date_to_ms(breakpoint_session["date"])
            prompt_needed = breakpoint_time > resolved_time

    n = len(pace_ratios)
    return RegimeShiftResult(
        breakpoints=result.breakpoints,
        promptNeeded=prompt_needed,
        cusumState={
            "upper": result.upperAccumulator[n - 1] if n > 0 else 0.0,
            "lower": result.lowerAccumulator[n - 1] if n > 0 else 0.0,
        },
    )
